subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestClassProcessor.java

```diff
@@ -83,21 +86,25 @@ public void stopNow() {
         @Override
         public void execute(String testClassName) {
             Class<?> klass = loadClass(testClassName);
-            if (isTopClass(klass)) {
-                testClasses.add(klass);
+            if (isInnerClassAnnotatedWithNested(klass) || isNestedClassInsideEnclosedRunner(klass)) {
+                return;
             }
+            testClasses.add(klass);
         }
 
         private void processAllTestClasses() {
             Launcher launcher = LauncherFactory.create();
             launcher.registerTestExecutionListeners(new JUnitPlatformTestExecutionListener(resultProcessor, clock, idGenerator, executionListener));
             launcher.execute(createLauncherDiscoveryRequest(testClasses));
         }
-
     }
 
-    private boolean isTopClass(Class<?> klass) {
-        return klass.getEnclosingClass() == null;
+    private boolean isInnerClassAnnotatedWithNested(Class<?> klass) {
+        if (klass.getEnclosingClass() == null) {
+            return false;
+        }
+        Annotation[] annotations = klass.getAnnotations();
+        return Stream.of(annotations).anyMatch(annotation -> "org.junit.jupiter.api.Nested".equals(annotation.annotationType().getName()));
```

Comment:
`@Nested` could be used as a meta-annotation. If at all possible, I still think it would be better to pass nested classes to the same worker instead of hard-coding special cases here. However, I don't know how much that would complicate the logic elsewhere.

Reply:
See my comment above - I don't think we should have logic for `@Nested` at all. Afaik the Junit platform console runner doesn't run *inner* classes by themselves, so Gradle shouldn't try to run them by themselves either. So instead of filtering by `@Nested`, I think we should ignore inner classes completely and only run top-level and static nested classes. Does that make sense?

Reply:
It does!